modules/tensor_mechanics/src/materials/ADLAROMANCEStressUpdateBase.C

```diff
@@ -794,8 +794,29 @@ ADLAROMANCEStressUpdateBase::computeStressFinalize(const ADRankTwoTensor & plast
   _cell_dislocations[_qp] = _old_input_values[_cell_output_index] + _cell_dislocation_increment;
   _wall_dislocations[_qp] = _old_input_values[_wall_output_index] + _wall_dislocation_increment;
 
-  for (unsigned int i = 0; i < _num_inputs; i++)
-    checkInputWindow(_input_values[i], _window_failure[i], _global_limits[i]);
```

Comment:
This skips the other input values no? Do the checking with the error correction above not work?

Reply:
In my print statements I was seeing that this line was checking the values input into the ROM while I would like to have a check for the output ROM values: this line would check the dislocations calculated at timestep (n-1) and I'd like to have a check on the dislocations calculated at timestep (n)

Reply:
Oh I see. I think you can just pass in output values, eg.

```
for (unsigned int i = 0; i < _num_outputs; i++)
    checkInputWindow(_output_values[i], _window_failure[i], _global_limits[i]);
```

Reply:
Yes, I agree that option does work; I did this in the first commit so that we can drop the second commit if the exception throwing isn't desired

Reply:
The main problem here is that `_global_limits` is really only applicable to the starting values of the input, and not the only value of the input. I had a similar check in there before the latest changes, but removed them based on suggestions from Aaron that there is no real lower limit. If you want to put `0.0` as the lower limit, that is probably fine:

```
for (unsigned int i = 0; i < _num_outputs; i++)
    checkInputWindow(_output_values[i], _window_failure[i], 0.0);
```

Reply:
> no real lower limit

Rather, the global limits are not the actual limits of the ROM.

Reply:
Are you suggesting that we remove the check on the ROM outputs at the end of the timestep all together then?

Reply:
This class is the only one that evolves the dislocation densities, so isn't it better do check if the calculated dislocations values from this timestep will be allowable as inputs in the next while we can still do something (e.g. cut the timestep) about it and try to get the simulation to run a bit further?

Reply:
> Are you suggesting that we remove the check on the ROM outputs at the end of the timestep all together then?

Yes, the confusing bit is that the rom (should) only develop dislocation densities that make sense. The parameters here are only valid for the initial dislocation densities, from which to evolve from, not what the absolute lowest limit is. We should try and have a call with Aaron and Laurent to clarify.

Reply:
>This class is the only one that evolves the dislocation densities, so isn't it better to check if the calculated dislocations values from this timestep will be allowable as inputs in the next while we can still do something (e.g. cut the timestep) about it and try to get the simulation to run a bit further?

There is no "valid lower limit" of densities determined by the ROM. I would agree that if they are unphysical (i.e. negative) then it should cut the timestep, hence my latest code snippet that should be used.

Reply:
Let's confirm, but I think the check you are adding here, and the check I have previously and you removed, are both unnecessary.

Reply:
I had thought, having been corrected previously last week by Aaron, that the initial condition values were the values we set with the `params.addRangeCheckedParam` and the evolution values were set by the `_global_limits`. Clarification on this from Aaron and Laurent would be useful

Reply:
> hence my latest code snippet that should be used
> ```
> for (unsigned int i = 0; i < _num_outputs; i++)
>    checkInputWindow(_output_values[i], _window_failure[i], 0.0);
>```
This won't work as written because there isn't a correspondence between the `_output_values` vector and the `_window_failure` vector. I have something similar using the index variables you created waiting in the wings. No sense in pushing it just to delete it

Reply:
Ah yes...well it would just be a hardcoding of the three output indexes then:
```
  checkInputWindow(_output_values[_cell_output_index], _window_failure[_cell_input_index], 0.0);
  checkInputWindow(_output_values[_wall_output_index], _window_failure[_wall_input_index], 0.0);
  checkInputWindow(_output_values[_strain_output_index], _window_failure[_stress_input_index], 0.0);
```

Reply:
Well...not totally either...but you get the idea. Basically, pass the values to `checkInputWindow` so you don't have to duplicate so much code.

Reply:
You don't need fuzzy on 0.0 (I think). Why not use `checkInputWindow`?

Reply:
>Why not use `checkInputWindow`?
Doesn't compile: it needs a vector not a constant.  I'm happy to continue discussing the best way to give the most useful information with the least amount of code once we hear from Laurent and Aaron, but if they say this check isn't necessary then this discussion is all wasted effort

Reply:
Okay, sounds like it was what I was thinking. I still think you are correct to have a sanity check on physical values for the dislocations here. I wonder if you'd just want to throw the exception, like you had previously, if one of the output values is unphysical:
```
  if (_cell_dislocations[_qp] < 0.0 || _wall_dislocations[_qp] < 0.0)
  {
    mooseExepction("In ",
                   _name,
                   ": Negative disclocation density calculated for (",
                   _cell_dislocations[_qp],
                   ") or wall (",
                   _wall_dislocations[_qp],
                   ").");
    _cell_dislocations[_qp] = _old_input_values[_cell_output_index];
    _wall_dislocations[_qp] = _old_input_values[_wall_output_index];
  }
```

Note the last bit of setting the dislocations to their old value is to prevent an unphysical value being passed else where and breaking another piece of code. I have seen this before with things like thermal conductivity that resulted in feedback loop that prevented a proper recovery via an exception. Yes, the dislocation densities are only used here, but it's good practice to protect future problems by allowing a graceful failure.

Reply:
I like it. If Aaron gives us the okay I'll add this in and repush